test/unit/communication/test_collective_async.cpp

```diff
@@ -0,0 +1,159 @@
+
+  if (root_rank == world.rank()) {
+    // use -> read
+    scheduleReduceRecvInPlace(ex_mpi, chain(), MPI_SUM, matrix(idx));
+
+    auto exp_tile = fixedValueTile(world.size() * (world.size() + 1) / 2);
+    CHECK_TILE_EQ(exp_tile, matrix.read(idx).get());
+  }
+  else {
+    // use -> read -> set -> read
+    scheduleReduceSend(ex_mpi, root_rank, chain(), MPI_SUM, matrix.read(idx));
+
+    CHECK_TILE_EQ(input_tile, matrix.read(idx).get());
+
+    auto new_tile = fixedValueTile(26);
+    matrix::test::set(matrix(idx).get(), new_tile);
+    CHECK_TILE_EQ(new_tile, matrix.read(idx).get());
+  }
+}
+
+TEST_F(CollectiveTest, ReduceInPlace) {
+  testReduceInPlace(world, newBlockMatrixContiguous<T, device>());
+  testReduceInPlace(world, newBlockMatrixStrided<T, device>());
+}
+
+template <class T, Device device>
+void testAllReduceInPlace(comm::Communicator world, matrix::Matrix<T, device> matrix) {
+  common::Pipeline<comm::Communicator> chain(world);
+
+  const auto ex_mpi = dlaf::getMPIExecutor<Backend::MC>();
+
+  const LocalTileIndex idx(0, 0);
+
+  // set -> use -> read
+  auto input_tile = fixedValueTile(world.rank() + 1);
+  matrix::test::set(matrix(idx).get(), input_tile);
+
+  auto after = scheduleAllReduceInPlace(ex_mpi, chain(), MPI_SUM, matrix(idx));
+
+  auto exp_tile = fixedValueTile(world.size() * (world.size() + 1) / 2);
+  CHECK_TILE_EQ(exp_tile, after.get());
+  CHECK_TILE_EQ(exp_tile, matrix.read(idx).get());
+}
+
+TEST_F(CollectiveTest, AllReduceInPlace) {
+  testAllReduceInPlace(world, newBlockMatrixContiguous<T, device>());
+  testAllReduceInPlace(world, newBlockMatrixStrided<T, device>());
+}
+
+template <class T, Device device>
+void testAllReduce(comm::Communicator world, matrix::Matrix<T, device> matA,
+                   matrix::Matrix<T, device> matB) {
+  common::Pipeline<comm::Communicator> chain(world);
+
+  const auto root_rank = world.size() - 1;
+  const auto ex_mpi = dlaf::getMPIExecutor<Backend::MC>();
+
+  const LocalTileIndex idx(0, 0);
+  matrix::Matrix<T, device>& mat_in = root_rank % 2 == 0 ? matA : matB;
+  matrix::Matrix<T, device>& mat_out = root_rank % 2 == 0 ? matB : matA;
+
+  // set -> use -> read
+  auto input_tile = fixedValueTile(world.rank() + 1);
+  matrix::test::set(mat_in(idx).get(), input_tile);
+
+  scheduleAllReduce(ex_mpi, chain(), MPI_SUM, mat_in.read(idx), mat_out(idx));
+
+  CHECK_TILE_EQ(input_tile, mat_in.read(idx).get());
+
+  auto exp_tile = fixedValueTile(world.size() * (world.size() + 1) / 2);
+  CHECK_TILE_EQ(exp_tile, mat_out.read(idx).get());
+}
+
+TEST_F(CollectiveTest, AllReduce) {
+  testAllReduce(world, newBlockMatrixContiguous<T, device>(), newBlockMatrixContiguous<T, device>());
+  testAllReduce(world, newBlockMatrixStrided<T, device>(), newBlockMatrixStrided<T, device>());
+  testAllReduce(world, newBlockMatrixContiguous<T, device>(), newBlockMatrixStrided<T, device>());
+}
+
+// TODO TEST AllReduce -> AllReduceInPlace Mixed
```

Comment:
@rasolca do we really want to test also mixed or is it enough like this?

Do you have any suggestion on how to improve tests? would you suggest to split them over different TEST_F so that we get more information in case of failure?

Reply:
About mixed tests I meant having ranks in the same communication in which part uses `allReduce` and the others `allReduceInPlace`. To me it looks superfluous, but you may have some edge cases it is worth checking.

Reply:
You are already mixing as you invert the matrices for odd ranks.

Reply:
In some sense, I agree, but in both cases the `scheduleAllReduceInPlace` is called, while I was thinking about mixing `scheduleAllReduceInPlace` and `scheduleAllReduce`.

Anyway, it is enough also for you, I would leave it as it is and remove the TODO.